Add missing chemical formulas; Fix sub/superscript display

redspah opened the pull request:
This PR makes no gameplay changes (hopefully), it's intended to be purely aesthetic / visual polish.

It adds formulas to most materials that didn't have it (doesn't alter their composition, just overrides the formulas), fixes the issues with subscript / superscript characters in formulas, and adjusts formulas of some substances where they were flat out incorrect (several monaline steps). It heavily makes use of the recently merged #4056.

==== Subscript / Superscript handling

A new set of custom characters (E010-E019) was added to fix issues with current superscript rendering, especially noticeable with Thorium 232. With the changes, the superscript characters look the same as subscript characters. Subscript 0 character is replaced with the new custom character E01A, because the previous one was very ugly for no reason.

Before/After:
![](https://i.imgur.com/VqN7m6N.png)


==== New/Changed Chemical Formulas


There are too many of them for me to list them all, so have a rundown of the possibly contentious ones. All of them are of course subject to change:

- Aludorite, Ceruclase, Orichalcum, Rubracium, Vulcanite and Vyroxeres are now space (`Sp`) Aluminum, Silver, Bismuth, Rubidium, Copper and Beryllium.
- Fiery Steel and Blood Iron use a new symbol, E004, representing the missing `🜏 U+17F0F ALCHEMICAL SYMBOL FOR BLACK SULFUR`. Mellion is changed correspondingly. 
![](https://i.imgur.com/i8gfqzE.png) 
- Tengam (and Charged Certus Quartz) use a new symbol, E00A, a higher quality version of `⚡ U+26A1 HIGH VOLTAGE SIGN`. All Magnetic elements also get a new Magnet symbol, E00E.
![](https://i.imgur.com/F1MjXaY.png)
- Deep Dark Iron uses a new symbol, E007, a higher quality version of `⛏ U+26CF PICK`
![](https://i.imgur.com/7lfpVlD.png)
- Dilithium:
![](https://i.imgur.com/zcCwScZ.png)
- Nether Star uses a new symbol, E005, a higher quality version of `⊗ U+2297 CIRCLED TIMES`
![](https://i.imgur.com/T8FDWBW.png)
- Unstable is now actually Iron divided by Diamond:
![](https://i.imgur.com/TimqTNK.png)
- Ordo / Perditio now use Logical OR and AND respectively, which fits in with the formulas for other elements added by @Vlamonster:
![](https://i.imgur.com/8etIJ6O.png)
- Blizz has a pretty snowflake :) (also affects Cryotheum)
![](https://i.imgur.com/v2x5mqq.png)
- White and Black Dwarf Matter now have funky symbols:
![](https://i.imgur.com/VCPUpyt.png)
- Mysterious Crystal uses a new symbol, E006, a higher quality version of `⍟ U+235F APL FUNCTIONAL SYMBOL CIRCLE STAR` 
![](https://i.imgur.com/YQVOEWu.png)
- Raw Stellar Plasma uses a new Galaxy symbol, E00C. MHDCSM parts are based on Raw Stellar Plasma, with the arrows to the sides conveying the "Constrained" aspect. 
![](https://i.imgur.com/M5gUhXB.png)
- Firestone got a new symbol, and Mellion got updated with it + Fiery Steel
![](https://i.imgur.com/hgEtVTo.png)
- Forcicium / Forcillum are _groovy_ now
![](https://i.imgur.com/gJzJTPn.png)
- Bastline and Monaline steps have adjusted formulas to more accurately reflect their products, step by step. Additionally, question mark subscript was added to denote compounds where a quantity of an element is unknown, as opposed to full size question mark which denotes an unknown elements. Many Monaline and Bastline steps use both.
![](https://i.imgur.com/MR2mpGC.png)
- Inspired by the fun flavor text of some naqfuel processing materials, naqfuels I-IV also get to have them now
![](https://i.imgur.com/tzQlKVE.png)

Edit: Updated with up-to-date tooltips.

RedSpah commented:
For posterity, here's a straightforward way to edit `glyph_sizes.bin`. I used [HxD](https://mh-nexus.de/en/hxd/) for this, but I imagine any other hex editor would be more than able to do the job well.

The process is straightforward: if you want to add/edit character `\uXXXX`, you have to edit the offset `XXXX` in the `glyph_sizes.bin` file. 

![](https://i.imgur.com/piIC3Ke.png)
*This PR adds characters `E010`-E01A` among others, so all of the accompanying offsets had to be edited*

At each offset / character, the upper nybble (left hex digit) represents the left x-coordinate, and the lower nybble (right hex digit) represents the right x-coordinate. Both of these are inclusive, so a value of `0F` would show the entire 16 pixel wide character, and `07` would only show its left half.

YannickMG commented:
The magnetic symbol is cute, but it seems a bit of a stretch for the idea of a "chemical formula". I'd definitely like to use that lightning symbol in UIs all over the pack but as an icon to stand for power information. I wonder if a magnet icon would be better suited for magnetism if you really want to use one... This is not a change request, just a comment and suggestion.

YannickMG commented:
> > The magnetic symbol is cute, but it seems a bit of a stretch for the idea of a "chemical formula". I'd definitely like to use that lightning symbol in UIs all over the pack but as an icon to stand for power information. I wonder if a magnet icon would be better suited for magnetism if you really want to use one. This is not a change request, just a comment and suggestion.
> 
> How does something like this look?
> 
> ![](https://camo.githubusercontent.com/834d6d89b326986e69fa63a90467a51958daacf652f515756d4234868163975f/68747470733a2f2f692e696d6775722e636f6d2f675a647a6e6e412e706e67)
> 
> And considering GTNH uses different formulas for normal and "solar grade" Silicon, I don't think magnetism is that much of a stretch to include in the formula.

That magnet looks very good, but also quite thick and large for a printed character, especially when you compare it to all the other symbols you introduced.

RedSpah commented:
> That magnet looks very good, but also quite thick and large for a printed character, especially when you compare it to all the other symbols you introduced.

The comment got me to tinker a bit, and realize how to make these look a bit better. This is another attempt: ![](https://i.imgur.com/6qOnKt3.png)

Basically, each custom character can be up to 16x16, but vanilla letters only take up to 10x14 of that space, and custom characters have to leave the two bottom rows empty to be level with vanilla text. the entire 16x16 space can be used for certain characters, but mixing those full size characters with regular text is gonna look weird.

![](https://i.imgur.com/QSTe1tN.png)

I'm gonna tinker with the characters I have to make them blend as well as possible

RedSpah commented:
Cleaned up the new glyphs so that they're all pixel-aligned, and resized them where appropriate. @Vlamonster sent me a few of her glyphs, and I added them to the PR after some tweaking, including the new glyph used for Raw Stellar Plasma / MHDCSM:

![](https://i.imgur.com/t7nIK1W.png)

Mobius0526 commented:
'Unstable' is actually 1/0 (iron ingots represent 1, diamonds represent 0), I suggest+∞ or Nan

FourIsTheNumber commented:
> 'Unstable' is actually 1/0 (iron ingots represent 1, diamonds represent 0), I suggest+∞ or Nan

I don't agree, the death message is even "player tried to divide by diamond". It's much funnier as Fe/C

Vlamonster commented:
Unstable might need a bigger left and right ceil/floor. It doesn't line up quite right now currently.

redspah commented:
With these changes done, is there anything still needing to be changed / corrected before this PR can be merged?

boubou19 commented:
Overall looks good, I have just a small nitpick: Could you change it so perditio's symbol doesn't look like Uranium's chemical symbol?

kamenskyi commented:
[redspah](https://github.com/redspah) the formulas : (C₁₄Os₁₁O₇Ag₃SpH₂O)₄?₁₀(Fs⚶)₆(⌘☯☯⌘)₅ 
 and ((Si₅O₁₀Fe)₃(Bi₂Te₃)₄ZrO₂Fe₅C)₅Og*Pr₁₅((C₁₄Os₁₁O₇Ag₃SpH₂O)₄?₁(Fs⚶)₆(⌘☯☯⌘)₅)₆〄₄ are missing an unchanged sign ₀
